rustls/src/client/mod.rs

```diff
@@ -662,9 +663,14 @@ impl ClientSession {
     /// we behave in the TLS protocol, `hostname` is the
     /// hostname of who we want to talk to.
     pub fn new(config: &Arc<ClientConfig>, hostname: webpki::DNSNameRef) -> ClientSession {
+        Self::try_new(config, hostname).unwrap()
```

Comment:
This function should be marked `#[deprecated]` or even better, removed. The documentation should explain why it is deprecated.

Reply:
You mean `new()`? Isn't it reasonable for many (types of) callers to continue to just panic for this type of failure? It seems like so far it really hasn't caused to many issues. I guess I could see the point that callers should be forced to make that decision for themselves.

Reply:
> You mean `new()`? Isn't it reasonable for many (types of) callers to continue to just panic for this type of failure? It seems like so far it really hasn't caused to many issues. I guess I could see the point that callers should be forced to make that decision for themselves.

Yes. To clarify, I think we should probably try to avoid using the CSPRNG in the `new` functions at all so that `try_new` isn't necessary; i.e. so `new()` becomes truly infallible. Faliing that, we should just remove the `new()` functions. Failing that, we should deprecated them and document why they are deprecated.

More generally, the API that Rustls should be designed so that Rustls will never panic itself. There might be cases where we have `unwrap()` in the code, but those should be unreachable code, and we should be confident that they are unreachable.

(I also think that eventually we should remove the `unwrap()`s from the example code.)

Reply:
I agree that we should only retain tightly audited `unwrap()` calls, but there quite a lot of `unwrap()` calls right now. I don't mind spending some time on this, but it would be a bit easier to do so if the work could be funded. @jsha would that be an option for your project?

Reply:
I'll ask!

Reply:
> I agree that we should only retain tightly audited `unwrap()` calls, but there quite a lot of `unwrap()` calls right now.

I filed issue #447 to track the effort of eliminating all the potential for panics within Rustls.